pkg/upload/js.go

```diff
@@ -108,6 +109,101 @@ func resolveSourceMapPaths(sourceMapPath string, outputPath string, logger log.L
 	return sourceMapPaths, nil
 }
 
+// Reads a JSON source map into memory. Required in order to check if the sourcesContent exists.
+func ReadSourceMap(path string, logger log.Logger) (map[string]interface{}, error) {
+	logger.Debug(fmt.Sprintf("Reading sourcemap %s", path))
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return map[string]interface{}{}, err
+	}
+	var sourceMapContents map[string]interface{}
+	err = json.Unmarshal(data, &sourceMapContents)
+	if err != nil {
+		return map[string]interface{}{}, err
+	}
+	return sourceMapContents, nil
+}
+
+func addSourcesContent(section map[string]interface{}, relativePath string, logger log.Logger) {
+	untypedSources, hasSources := section["sources"]
+	if !hasSources {
+		logger.Warn("Source map doesn't have required sources field")
+		return
+	}
+	sources, sourcesValid := untypedSources.([]interface{})
+	if !sourcesValid {
+		logger.Warn("Sources aren't an array")
+		return
+	}
+
+	// Skip if the sources content is already valid
+	untypedSourcesContent, hasSourcesContent := section["sourcesContent"]
+	if hasSourcesContent {
+		sourcesContent, sourcesContentValid := untypedSourcesContent.([]interface{})
+		if sourcesContentValid && len(sourcesContent) == len(sources) {
+			logger.Debug("SourcesContent is already populated")
+			return
+		}
+	}
```

Comment:
Where is this untyped and valid content logic taken from? I couldn't glean it from the old tool or any online spec. Would be good to see a summary of the expected file format, or citation, and/or some more explanatory comments for future readers.

Reply:
This logic is based on the JS implementation (but that obviously didn't have types) as well as the sourcemap specification https://bit.ly/sourcemap which dictates the types. I have added in a comment at the top of that function